haskell.lib.buildStackProject: allow passing in a custom GHC version.

mboes opened the pull request:
Previously, the user could only compile using the default version of GHC
in Nixpkgs. Now this can be changed by setting the `ghc` attribute
appropriately.

mboes committed:
haskell.lib.buildStackProject: allow passing in a custom GHC version.

Previously, the user could only compile using the default version of
GHC in Nixpkgs. Now this can be changed by setting the `ghc` attribute
appropriately.

mboes committed:
doc: Document how to set custom GHC version in haskell.lib.buildStackProject.

YPares commented:
Given this recent addition to stack https://github.com/commercialhaskell/stack/commit/4edd98544b05f3e77dcd70feed13693513b7c322
I believe there are two options:
- either use this PR as and modify stack's doc to switch from `buildInputs = [ ghc glpk pcre ];`
to `inherit ghc; buildInputs = [ glpk pcre ];`
- modify this PR so `generic-stack-builder` doesn't handle ghc specifically, as it will just be part of the `buildInputs` (this is how ghc is handled by Stack/Nix right when we give to Stack a list of packages instead of a full `shell.nix`: ghc is just added to the package list before creating the nix expression).

obadz commented:
cc @peti @bennofs

mboes commented:
Option (2) breaks backwards compatibility. Which is why I did it the other way. And also for consistency: `buildPythonPackage` et al similarly handle the compiler/runtime choice internally. It's not something the user should be adding to `buildInputs` herself.

YPares commented:
@m Okay then,
on Stack front, everything is ready. We are waiting for this PR to be merged, given that now it's necessary for us to be able to pass a GHC derivation to the shell.nix file

mboes committed:
haskell.lib.buildStackProject: fix treatment of LD_LIBRARY_PATH.

Don't make LD_LIBRARY_PATH pick up cwd when original string is empty,
due to colon terminated search path.

mboes commented:
@peti all Stack projects tracking current LTS and using `haskell.lib.buildStackProject` are essentially broken until this PR is merged, due to the default compiler changing to GHC 8 in recent Nixpkgs.

peti commented on doc/languages-frameworks/haskell.md:
```diff
@@ -378,6 +378,23 @@ special options turned on:
+
+    let R = pkgs.R.override { enableStrictBarrier = true; };
+    in
+	haskell.lib.buildStackProject {
```
This looks like the code snippet is not properly indented. Did you check whether the generated manual looks okay?

mboes replied:
fixed.

mboes committed:
doc(haskell): Remove pesky tabs in code snippets.
